File: auxiliaryjar/src/main/java/com/zhf/auxiliaryjar/verification_code/WiseEditText.java
package com.zhf.auxiliaryjar.verification_code;

import android.content.Context;
import android.support.v7.widget.AppCompatEditText;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputConnection;
import android.view.inputmethod.InputConnectionWrapper;


public class WiseEditText extends AppCompatEditText {


    private OnKeyListener keyListener;

    public WiseEditText(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public WiseEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public WiseEditText(Context context) {
        super(context);
    }

    @Override
    public InputConnection onCreateInputConnection(EditorInfo outAttrs) {
        return new MyInputConnection(super.onCreateInputConnection(outAttrs),
                true);
    }

    private class MyInputConnection extends InputConnectionWrapper {

        public MyInputConnection(InputConnection target, boolean mutable) {
            super(target, mutable);
        }

        @Override
        public boolean sendKeyEvent(KeyEvent event) {
            if (keyListener != null) {
                keyListener.onKey(WiseEditText.this,event.getKeyCode(),event);
            }
            return super.sendKeyEvent(event);
        }

        @Override
        public boolean deleteSurroundingText(int beforeLength, int afterLength) {
            // magic: in latest Android, deleteSurroundingText(1, 0) will be called for backspace
            if (beforeLength == 1 && afterLength == 0) {
                // backspace
                return sendKeyEvent(new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DEL))
                        && sendKeyEvent(new KeyEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_DEL));
            }

            return super.deleteSurroundingText(beforeLength, afterLength);
        }

    }

    //设置监听回调
    public void setSoftKeyListener(OnKeyListener listener){
        keyListener = listener;
    }

}

File: auxiliaryjar/src/main/java/com/zhf/auxiliaryjar/verification_code/WiseEditText.kt
package com.zhf.auxiliaryjar.verification_code

import android.content.Context
import android.support.v7.widget.AppCompatEditText
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputConnection
import android.view.inputmethod.InputConnectionWrapper


class WiseEditText : AppCompatEditText {


	private var keyListener: OnKeyListener? = null

	constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {}

	constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

	constructor(context: Context) : super(context) {}

	override fun onCreateInputConnection(outAttrs: EditorInfo): InputConnection {
		return MyInputConnection(super.onCreateInputConnection(outAttrs),
				true)
	}

	private inner class MyInputConnection(target: InputConnection, mutable: Boolean) : InputConnectionWrapper(target, mutable) {

		override fun sendKeyEvent(event: KeyEvent): Boolean {
			if (keyListener != null) {
				keyListener!!.onKey(this@WiseEditText, event.keyCode, event)
			}
			return super.sendKeyEvent(event)
		}

		override fun deleteSurroundingText(beforeLength: Int, afterLength: Int): Boolean {
			// magic: in latest Android, deleteSurroundingText(1, 0) will be called for backspace
			return if (beforeLength == 1 && afterLength == 0) {
				// backspace
				sendKeyEvent(KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DEL)) && sendKeyEvent(KeyEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_DEL))
			} else super.deleteSurroundingText(beforeLength, afterLength)

		}

	}

	//设置监听回调
	fun setSoftKeyListener(listener: OnKeyListener) {
		keyListener = listener
	}

}